[java] Maven  PMD plugin fails to process some files without any explanation

Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core].

**Affects PMD Version:** 
6.0.1
**Rule:**
No specific rule - just fails
**Description:**
I am trying to upgrade from 5.6.1 to 6.0.1 and it fails where in the past it succeeded. All it says is:

>> Caused by: java.lang.RuntimeException: org.apache.maven.reporting.MavenReportException: Found 2 PMD processing errors

**Code Sample demonstrating the issue:**

Easily reproducible - simply

* Clone [MINA SSHD](https://github.com/apache/mina-sshd/)
* Edit its _pom.xml_ and replace `pmd.version` property with _6.0.1_
* Run `mvn clean install`

This is one of the 2 failures - `BasePath` (the other one is `JaasPasswordAuthenticatorTest`)

```java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.sshd.common.file.util;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.nio.file.FileSystem;
import java.nio.file.Path;
import java.nio.file.ProviderMismatchException;
import java.nio.file.WatchEvent;
import java.nio.file.WatchKey;
import java.nio.file.WatchService;
import java.util.AbstractList;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Deque;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;

import org.apache.sshd.common.util.GenericUtils;

public abstract class BasePath<T extends BasePath<T, FS>, FS extends BaseFileSystem<T>> implements Path {

    protected final String root;
    protected final List<String> names;
    private final FS fileSystem;
    private String strValue;
    private int hashValue;

    public BasePath(FS fileSystem, String root, List<String> names) {
        this.fileSystem = Objects.requireNonNull(fileSystem, "No file system provided");
        this.root = root;
        this.names = names;
    }

    @SuppressWarnings("unchecked")
    protected T asT() {
        return (T) this;
    }

    protected T create(String root, String... names) {
        return create(root, GenericUtils.unmodifiableList(names));
    }

    protected T create(String root, Collection<String> names) {
        return create(root, GenericUtils.unmodifiableList(names));
    }

    protected T create(String root, List<String> names) {
        return fileSystem.create(root, names);
    }

    @Override
    public FS getFileSystem() {
        return fileSystem;
    }

    @Override
    public boolean isAbsolute() {
        return root != null;
    }

    @Override
    public T getRoot() {
        if (isAbsolute()) {
            return create(root);
        }
        return null;
    }

    @Override
    public T getFileName() {
        if (!names.isEmpty()) {
            return create(null, names.get(names.size() - 1));
        }
        return null;
    }

    @Override
    public T getParent() {
        if (names.isEmpty() || ((names.size() == 1) && (root == null))) {
            return null;
        }
        return create(root, names.subList(0, names.size() - 1));
    }

    @Override
    public int getNameCount() {
        return names.size();
    }

    @Override
    public T getName(int index) {
        int maxIndex = getNameCount();
        if ((index < 0) || (index >= maxIndex)) {
            throw new IllegalArgumentException("Invalid name index " + index + " - not in range [0-" + maxIndex + "]");
        }
        return create(null, names.subList(index, index + 1));
    }

    @Override
    public T subpath(int beginIndex, int endIndex) {
        int maxIndex = getNameCount();
        if ((beginIndex < 0) || (beginIndex >= maxIndex) || (endIndex > maxIndex) || (beginIndex >= endIndex)) {
            throw new IllegalArgumentException("subpath(" + beginIndex + "," + endIndex + ") bad index range - allowed [0-" + maxIndex + "]");
        }
        return create(null, names.subList(beginIndex, endIndex));
    }

    protected boolean startsWith(List<?> list, List<?> other) {
        return list.size() >= other.size() && list.subList(0, other.size()).equals(other);
    }

    @Override
    public boolean startsWith(Path other) {
        T p1 = asT();
        T p2 = checkPath(other);
        return Objects.equals(p1.getFileSystem(), p2.getFileSystem())
                && Objects.equals(p1.root, p2.root)
                && startsWith(p1.names, p2.names);
    }

    @Override
    public boolean startsWith(String other) {
        return startsWith(getFileSystem().getPath(other));
    }

    protected boolean endsWith(List<?> list, List<?> other) {
        return other.size() <= list.size() && list.subList(list.size() - other.size(), list.size()).equals(other);
    }

    @Override
    public boolean endsWith(Path other) {
        T p1 = asT();
        T p2 = checkPath(other);
        if (p2.isAbsolute()) {
            return p1.compareTo(p2) == 0;
        }
        return endsWith(p1.names, p2.names);
    }

    @Override
    public boolean endsWith(String other) {
        return endsWith(getFileSystem().getPath(other));
    }

    protected boolean isNormal() {
        int count = getNameCount();
        if ((count == 0) || ((count == 1) && !isAbsolute())) {
            return true;
        }
        boolean foundNonParentName = isAbsolute(); // if there's a root, the path doesn't start with ..
        boolean normal = true;
        for (String name : names) {
            if (name.equals("..")) {
                if (foundNonParentName) {
                    normal = false;
                    break;
                }
            } else {
                if (name.equals(".")) {
                    normal = false;
                    break;
                }
                foundNonParentName = true;
            }
        }
        return normal;
    }

    @Override
    public T normalize() {
        if (isNormal()) {
            return asT();
        }

        Deque<String> newNames = new ArrayDeque<>();
        for (String name : names) {
            if (name.equals("..")) {
                String lastName = newNames.peekLast();
                if (lastName != null && !lastName.equals("..")) {
                    newNames.removeLast();
                } else if (!isAbsolute()) {
                    // if there's a root and we have an extra ".." that would go up above the root, ignore it
                    newNames.add(name);
                }
            } else if (!name.equals(".")) {
                newNames.add(name);
            }
        }

        return newNames.equals(names) ? asT() : create(root, newNames);
    }

    @Override
    public T resolve(Path other) {
        T p1 = asT();
        T p2 = checkPath(other);
        if (p2.isAbsolute()) {
            return p2;
        }
        if (p2.names.isEmpty()) {
            return p1;
        }
        String[] names = new String[p1.names.size() + p2.names.size()];
        int index = 0;
        for (String p : p1.names) {
            names[index++] = p;
        }
        for (String p : p2.names) {
            names[index++] = p;
        }
        return create(p1.root, names);
    }

    @Override
    public T resolve(String other) {
        return resolve(getFileSystem().getPath(other, GenericUtils.EMPTY_STRING_ARRAY));
    }

    @Override
    public Path resolveSibling(Path other) {
        Objects.requireNonNull(other, "Missing sibling path argument");
        T parent = getParent();
        return parent == null ? other : parent.resolve(other);
    }

    @Override
    public Path resolveSibling(String other) {
        return resolveSibling(getFileSystem().getPath(other, GenericUtils.EMPTY_STRING_ARRAY));
    }

    @Override
    public T relativize(Path other) {
        T p1 = asT();
        T p2 = checkPath(other);
        if (!Objects.equals(p1.getRoot(), p2.getRoot())) {
            throw new IllegalArgumentException("Paths have different roots: " + this + ", " + other);
        }
        if (p2.equals(p1)) {
            return create(null);
        }
        if (p1.root == null && p1.names.isEmpty()) {
            return p2;
        }
        // Common subsequence
        int sharedSubsequenceLength = 0;
        for (int i = 0; i < Math.min(p1.names.size(), p2.names.size()); i++) {
            if (p1.names.get(i).equals(p2.names.get(i))) {
                sharedSubsequenceLength++;
            } else {
                break;
            }
        }
        int extraNamesInThis = Math.max(0, p1.names.size() - sharedSubsequenceLength);
        List<String> extraNamesInOther = (p2.names.size() <= sharedSubsequenceLength)
                ? Collections.emptyList()
                : p2.names.subList(sharedSubsequenceLength, p2.names.size());
        List<String> parts = new ArrayList<>(extraNamesInThis + extraNamesInOther.size());
        // add .. for each extra name in this path
        parts.addAll(Collections.nCopies(extraNamesInThis, ".."));
        // add each extra name in the other path
        parts.addAll(extraNamesInOther);
        return create(null, parts);
    }

    @Override
    public T toAbsolutePath() {
        if (isAbsolute()) {
            return asT();
        }
        return fileSystem.getDefaultDir().resolve(this);
    }

    @Override
    public URI toUri() {
        File file = toFile();
        return file.toURI();
    }

    @Override
    public File toFile() {
        throw new UnsupportedOperationException("To file " + toAbsolutePath() + " N/A");
    }

    @Override
    public WatchKey register(WatchService watcher, WatchEvent.Kind<?>... events) throws IOException {
        return register(watcher, events, (WatchEvent.Modifier[]) null);
    }

    @Override
    public WatchKey register(WatchService watcher, WatchEvent.Kind<?>[] events, WatchEvent.Modifier... modifiers) throws IOException {
        throw new UnsupportedOperationException("Register to watch " + toAbsolutePath() + " N/A");
    }

    @Override
    public Iterator<Path> iterator() {
        return new AbstractList<Path>() {
            @Override
            public Path get(int index) {
                return getName(index);
            }

            @Override
            public int size() {
                return getNameCount();
            }
        }.iterator();
    }

    @Override
    public int compareTo(Path paramPath) {
        T p1 = asT();
        T p2 = checkPath(paramPath);
        int c = compare(p1.root, p2.root);
        if (c != 0) {
            return c;
        }
        for (int i = 0; i < Math.min(p1.names.size(), p2.names.size()); i++) {
            String n1 = p1.names.get(i);
            String n2 = p2.names.get(i);
            c = compare(n1, n2);
            if (c != 0) {
                return c;
            }
        }
        return p1.names.size() - p2.names.size();
    }

    protected int compare(String s1, String s2) {
        if (s1 == null) {
            return s2 == null ? 0 : -1;
        } else {
            return s2 == null ? +1 : s1.compareTo(s2);
        }
    }

    @SuppressWarnings("unchecked")
    protected T checkPath(Path paramPath) {
        Objects.requireNonNull(paramPath, "Missing path argument");
        if (paramPath.getClass() != getClass()) {
            throw new ProviderMismatchException("Path is not of this class: " + paramPath + "[" + paramPath.getClass().getSimpleName() + "]");
        }
        T t = (T) paramPath;

        FileSystem fs = t.getFileSystem();
        if (fs.provider() != this.fileSystem.provider()) {
            throw new ProviderMismatchException("Mismatched providers for " + t);
        }
        return t;
    }

    @Override
    public int hashCode() {
        synchronized (this) {
            if (hashValue == 0) {
                hashValue = calculatedHashCode();
                if (hashValue == 0) {
                    hashValue = 1;
                }
            }
        }

        return hashValue;
    }

    protected int calculatedHashCode() {
        return Objects.hash(getFileSystem(), root, names);
    }

    @Override
    public boolean equals(Object obj) {
        return (obj instanceof Path) && (compareTo((Path) obj) == 0);
    }

    @Override
    public String toString() {
        synchronized (this) {
            if (strValue == null) {
                strValue = asString();
            }
        }

        return strValue;
    }

    protected String asString() {
        StringBuilder sb = new StringBuilder();
        if (root != null) {
            sb.append(root);
        }

        String separator = getFileSystem().getSeparator();
        for (String name : names) {
            if ((sb.length() > 0) && (sb.charAt(sb.length() - 1) != '/')) {
                sb.append(separator);
            }
            sb.append(name);
        }

        return sb.toString();
    }
}
```

**Running PMD through:** *[Maven]*
 
Running the build with _-X_ flag yields no useful information (at least not to me):

```
[INFO] >>> maven-pmd-plugin:3.8:check (pmd-checker) > :pmd @ sshd-core >>>
[DEBUG] Dependency collection stats: {ConflictMarker.analyzeTime=46575, ConflictMarker.markTime=33755, ConflictMarker.nodeCount=37, ConflictIdSorter.graphTime=41020, ConflictIdSorter.topsortTime=22645, ConflictIdSorter.conflictIdCount=28, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=348664, ConflictResolver.conflictItemCount=36, DefaultDependencyCollector.collectTime=2950819, DefaultDependencyCollector.transformTime=509749}
[DEBUG] org.apache.sshd:sshd-core:jar:1.7.1-SNAPSHOT
[DEBUG]    org.slf4j:slf4j-api:jar:1.7.25:compile
[DEBUG]    org.apache.mina:mina-core:jar:2.0.16:compile (optional)
[DEBUG]    tomcat:tomcat-apr:jar:5.5.23:compile (optional)
[DEBUG]    org.bouncycastle:bcpg-jdk15on:jar:1.59:compile (optional)
[DEBUG]       org.bouncycastle:bcprov-jdk15on:jar:1.59:compile (version managed from 1.59) (optional)
[DEBUG]    org.bouncycastle:bcpkix-jdk15on:jar:1.59:compile (optional)
[DEBUG]    net.i2p.crypto:eddsa:jar:0.2.0:compile (optional)
[DEBUG]    org.slf4j:jcl-over-slf4j:jar:1.7.25:test
[DEBUG]    org.slf4j:slf4j-log4j12:jar:1.7.25:test
[DEBUG]       log4j:log4j:jar:1.2.17:test (version managed from 1.2.17) (exclusions managed)
[DEBUG]    com.jcraft:jsch:jar:0.1.54:test
[DEBUG]    com.jcraft:jzlib:jar:1.1.3:test
[DEBUG]    org.springframework:spring-context:jar:5.0.3.RELEASE:test
[DEBUG]       org.springframework:spring-aop:jar:5.0.3.RELEASE:test (version managed from 5.0.3.RELEASE)
[DEBUG]       org.springframework:spring-beans:jar:5.0.3.RELEASE:test (version managed from 5.0.3.RELEASE)
[DEBUG]       org.springframework:spring-core:jar:5.0.3.RELEASE:test (version managed from 5.0.3.RELEASE)
[DEBUG]          org.springframework:spring-jcl:jar:5.0.3.RELEASE:test (version managed from 5.0.3.RELEASE)
[DEBUG]       org.springframework:spring-expression:jar:5.0.3.RELEASE:test (version managed from 5.0.3.RELEASE)
[DEBUG]    junit:junit:jar:4.12:test
[DEBUG]       org.hamcrest:hamcrest-core:jar:1.3:test
[DEBUG]    org.mockito:mockito-core:jar:2.13.0:test
[DEBUG]       net.bytebuddy:byte-buddy:jar:1.7.9:test
[DEBUG]       net.bytebuddy:byte-buddy-agent:jar:1.7.9:test
[DEBUG]       org.objenesis:objenesis:jar:2.6:test
[DEBUG]    commons-httpclient:commons-httpclient:jar:3.1:test (exclusions managed)
[DEBUG]       commons-codec:commons-codec:jar:1.2:test
[DEBUG]    ch.ethz.ganymed:ganymed-ssh2:jar:262:test
[DEBUG]    org.apache.servicemix.bundles:org.apache.servicemix.bundles.not-yet-commons-ssl:jar:0.3.11_1:test
[INFO] 
[INFO] --- maven-pmd-plugin:3.8:pmd (pmd) @ sshd-core ---
[DEBUG] Configuring mojo org.apache.maven.plugins:maven-pmd-plugin:3.8:pmd from plugin realm ClassRealm[plugin>org.apache.maven.plugins:maven-pmd-plugin:3.8, parent: sun.misc.Launcher$AppClassLoader@33909752]
[DEBUG] Configuring mojo 'org.apache.maven.plugins:maven-pmd-plugin:3.8:pmd' with basic configurator -->
[DEBUG]   (f) aggregate = false
[DEBUG]   (f) analysisCache = false
[DEBUG]   (f) analysisCacheLocation = D:\Projects\apache\mina-sshd\sshd-core\target/pmd/pmd.cache
[DEBUG]   (f) benchmark = false
[DEBUG]   (f) benchmarkOutputFilename = D:\Projects\apache\mina-sshd\sshd-core\target/pmd-benchmark.txt
[DEBUG]   (f) compileSourceRoots = [D:\Projects\apache\mina-sshd\sshd-core\src\main\java, D:\Projects\apache\mina-sshd\sshd-core\target\generated-sources\annotations]
[DEBUG]   (f) excludeRoots = [D:\Projects\apache\mina-sshd\sshd-core\target\generated-sources\java]
[DEBUG]   (f) format = xml
[DEBUG]   (f) includeTests = true
[DEBUG]   (f) includeXmlInSite = false
[DEBUG]   (f) inputEncoding = UTF-8
[DEBUG]   (f) language = java
[DEBUG]   (f) linkXRef = true
[DEBUG]   (f) minimumPriority = 5
[DEBUG]   (f) outputDirectory = D:\Projects\apache\mina-sshd\sshd-core\target\site
[DEBUG]   (f) outputEncoding = UTF-8
[DEBUG]   (f) project = MavenProject: org.apache.sshd:sshd-core:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\sshd-core\pom.xml
[DEBUG]   (f) reactorProjects = [MavenProject: org.apache.sshd:sshd:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\pom.xml, MavenProject: org.apache.sshd:sshd-core:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\sshd-core\pom.xml, MavenProject: org.apache.sshd:sshd-ldap:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\sshd-ldap\pom.xml, MavenProject: org.apache.sshd:sshd-git:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\sshd-git\pom.xml, MavenProject: org.apache.sshd:sshd-contrib:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\sshd-contrib\pom.xml, MavenProject: org.apache.sshd:sshd-spring-sftp:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\sshd-spring-sftp\pom.xml, MavenProject: org.apache.sshd:apache-sshd:1.7.1-SNAPSHOT @ D:\Projects\apache\mina-sshd\assembly\pom.xml]
[DEBUG]   (s) rulesets = [D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml]
[DEBUG]   (f) skip = false
[DEBUG]   (f) skipEmptyReport = true
[DEBUG]   (f) skipPmdError = false
[DEBUG]   (f) sourceEncoding = UTF-8
[DEBUG]   (f) targetDirectory = D:\Projects\apache\mina-sshd\sshd-core\target
[DEBUG]   (f) targetJdk = 1.8
[DEBUG]   (f) testSourceRoots = [D:\Projects\apache\mina-sshd\sshd-core\src\test\java, D:\Projects\apache\mina-sshd\sshd-core\target\generated-test-sources\test-annotations]
[DEBUG]   (f) typeResolution = true
[DEBUG]   (f) xrefLocation = D:\Projects\apache\mina-sshd\sshd-core\target\site\xref
[DEBUG]   (f) xrefTestLocation = D:\Projects\apache\mina-sshd\sshd-core\target\site\xref-test
[DEBUG] -- end configuration --
[DEBUG] Using language LanguageModule:Java(JavaLanguageModule)+version:1.8
[DEBUG] Using aux classpath: [D:\Projects\apache\mina-sshd\sshd-core\target\test-classes, D:\Projects\apache\mina-sshd\sshd-core\target\classes, D:\Projects\repository\org\slf4j\slf4j-api\1.7.25\slf4j-api-1.7.25.jar, D:\Projects\repository\org\apache\mina\mina-core\2.0.16\mina-core-2.0.16.jar, D:\Projects\repository\tomcat\tomcat-apr\5.5.23\tomcat-apr-5.5.23.jar, D:\Projects\repository\org\bouncycastle\bcpg-jdk15on\1.59\bcpg-jdk15on-1.59.jar, D:\Projects\repository\org\bouncycastle\bcprov-jdk15on\1.59\bcprov-jdk15on-1.59.jar, D:\Projects\repository\org\bouncycastle\bcpkix-jdk15on\1.59\bcpkix-jdk15on-1.59.jar, D:\Projects\repository\net\i2p\crypto\eddsa\0.2.0\eddsa-0.2.0.jar, D:\Projects\repository\org\slf4j\jcl-over-slf4j\1.7.25\jcl-over-slf4j-1.7.25.jar, D:\Projects\repository\org\slf4j\slf4j-log4j12\1.7.25\slf4j-log4j12-1.7.25.jar, D:\Projects\repository\log4j\log4j\1.2.17\log4j-1.2.17.jar, D:\Projects\repository\com\jcraft\jsch\0.1.54\jsch-0.1.54.jar, D:\Projects\repository\com\jcraft\jzlib\1.1.3\jzlib-1.1.3.jar, D:\Projects\repository\org\springframework\spring-context\5.0.3.RELEASE\spring-context-5.0.3.RELEASE.jar, D:\Projects\repository\org\springframework\spring-aop\5.0.3.RELEASE\spring-aop-5.0.3.RELEASE.jar, D:\Projects\repository\org\springframework\spring-beans\5.0.3.RELEASE\spring-beans-5.0.3.RELEASE.jar, D:\Projects\repository\org\springframework\spring-core\5.0.3.RELEASE\spring-core-5.0.3.RELEASE.jar, D:\Projects\repository\org\springframework\spring-jcl\5.0.3.RELEASE\spring-jcl-5.0.3.RELEASE.jar, D:\Projects\repository\org\springframework\spring-expression\5.0.3.RELEASE\spring-expression-5.0.3.RELEASE.jar, D:\Projects\repository\junit\junit\4.12\junit-4.12.jar, D:\Projects\repository\org\hamcrest\hamcrest-core\1.3\hamcrest-core-1.3.jar, D:\Projects\repository\org\mockito\mockito-core\2.13.0\mockito-core-2.13.0.jar, D:\Projects\repository\net\bytebuddy\byte-buddy\1.7.9\byte-buddy-1.7.9.jar, D:\Projects\repository\net\bytebuddy\byte-buddy-agent\1.7.9\byte-buddy-agent-1.7.9.jar, D:\Projects\repository\org\objenesis\objenesis\2.6\objenesis-2.6.jar, D:\Projects\repository\commons-httpclient\commons-httpclient\3.1\commons-httpclient-3.1.jar, D:\Projects\repository\commons-codec\commons-codec\1.2\commons-codec-1.2.jar, D:\Projects\repository\ch\ethz\ganymed\ganymed-ssh2\262\ganymed-ssh2-262.jar, D:\Projects\repository\org\apache\servicemix\bundles\org.apache.servicemix.bundles.not-yet-commons-ssl\0.3.11_1\org.apache.servicemix.bundles.not-yet-commons-ssl-0.3.11_1.jar]
[DEBUG] Preparing ruleset: D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml
[DEBUG] Before: D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml After: sshd-pmd-ruleset.xml
[DEBUG] URLResourceLoader: Exception when looking for 'D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml' at ''
java.net.MalformedURLException: unknown protocol: d
    at java.net.URL.<init> (URL.java:600)
    at java.net.URL.<init> (URL.java:490)
    at java.net.URL.<init> (URL.java:439)
    at org.codehaus.plexus.resource.loader.URLResourceLoader.getResource (URLResourceLoader.java:71)
    at org.codehaus.plexus.resource.DefaultResourceManager.getResource (DefaultResourceManager.java:159)
    at org.codehaus.plexus.resource.DefaultResourceManager.getResourceAsFile (DefaultResourceManager.java:91)
    at org.apache.maven.plugin.pmd.PmdReport.executePmd (PmdReport.java:354)
    at org.apache.maven.plugin.pmd.PmdReport.executePmdWithClassloader (PmdReport.java:311)
    at org.apache.maven.plugin.pmd.PmdReport.canGenerateReport (PmdReport.java:285)
    at org.apache.maven.reporting.AbstractMavenReport.execute (AbstractMavenReport.java:119)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions (MojoExecutor.java:353)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:198)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[DEBUG] URLResourceLoader: Exception when looking for 'D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml
java.net.MalformedURLException: unknown protocol: d
    at java.net.URL.<init> (URL.java:600)
    at java.net.URL.<init> (URL.java:490)
    at java.net.URL.<init> (URL.java:439)
    at org.codehaus.plexus.resource.loader.URLResourceLoader.getResource (URLResourceLoader.java:123)
    at org.codehaus.plexus.resource.DefaultResourceManager.getResource (DefaultResourceManager.java:159)
    at org.codehaus.plexus.resource.DefaultResourceManager.getResourceAsFile (DefaultResourceManager.java:91)
    at org.apache.maven.plugin.pmd.PmdReport.executePmd (PmdReport.java:354)
    at org.apache.maven.plugin.pmd.PmdReport.executePmdWithClassloader (PmdReport.java:311)
    at org.apache.maven.plugin.pmd.PmdReport.canGenerateReport (PmdReport.java:285)
    at org.apache.maven.reporting.AbstractMavenReport.execute (AbstractMavenReport.java:119)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions (MojoExecutor.java:353)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:198)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[DEBUG] The resource 'D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml' was not found with resourceLoader org.codehaus.plexus.resource.loader.URLResourceLoader.
[DEBUG] The resource 'D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml' was found as D:\Projects\apache\mina-sshd\sshd-pmd-ruleset.xml.
[DEBUG] Exclusions: **/*~,**/#*#,**/.#*,**/%*%,**/._*,**/CVS,**/CVS/**,**/.cvsignore,**/RCS,**/RCS/**,**/SCCS,**/SCCS/**,**/vssver.scc,**/project.pj,**/.svn,**/.svn/**,**/.arch-ids,**/.arch-ids/**,**/.bzr,**/.bzr/**,**/.MySCMServerInfo,**/.DS_Store,**/.metadata,**/.metadata/**,**/.hg,**/.hgignore,**/.hg/**,**/.git,**/.gitignore,**/.gitattributes,**/.git/**,**/BitKeeper,**/BitKeeper/**,**/ChangeSet,**/ChangeSet/**,**/_darcs,**/_darcs/**,**/.darcsrepo,**/.darcsrepo/**,**/-darcs-backup*,**/.darcs-temp-mail
[DEBUG] Inclusions: **/*.java
[DEBUG] Searching for files in directory D:\Projects\apache\mina-sshd\sshd-core\src\main\java
[DEBUG] Searching for files in directory D:\Projects\apache\mina-sshd\sshd-core\target\generated-sources\annotations
[DEBUG] Searching for files in directory D:\Projects\apache\mina-sshd\sshd-core\src\test\java
[DEBUG] Searching for files in directory D:\Projects\apache\mina-sshd\sshd-core\target\generated-test-sources\test-annotations
[DEBUG] Executing PMD...
[DEBUG] PMD finished. Found 0 violations.
[ERROR] PMD processing errors:
[ERROR] D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java: Error while processing D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java
D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java: Error while parsing D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Apache Mina SSHD ................................... SUCCESS [  7.673 s]
[INFO] Apache Mina SSHD :: Core ........................... FAILURE [01:22 min]
[INFO] Apache Mina SSHD :: LDAP ........................... SKIPPED
[INFO] Apache Mina SSHD :: Git ............................ SKIPPED
[INFO] Apache Mina SSHD :: Contributions .................. SKIPPED
[INFO] Apache Mina SSHD :: Spring integration SFTP adapter  SKIPPED
[INFO] Apache Mina SSHD :: Assembly ....................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:31 min
[INFO] Finished at: 2018-01-31T21:48:26+02:00
[INFO] Final Memory: 82M/794M
[INFO] ------------------------------------------------------------------------
```

Still happens after upgrade to PMD plugin 3.9.0

```
[INFO] >>> maven-pmd-plugin:3.9.0:check (pmd-checker) > :pmd @ sshd-core >>>
[INFO]
[INFO] --- maven-pmd-plugin:3.9.0:pmd (pmd) @ sshd-core ---
[WARNING] This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.0.1/pmd_userdocs_getting_started.html#incremental-analysis
[WARNING] The RuleSet rulesets/java/braces.xml has been deprecated and will be removed in PMD 7.0.0
[WARNING] The RuleSet rulesets/java/empty.xml has been deprecated and will be removed in PMD 7.0.0
[WARNING] The RuleSet rulesets/java/finalizers.xml has been deprecated and will be removed in PMD 7.0.0
[ERROR] PMD processing errors:
[ERROR] D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java: Error while processing D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java
D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java: Error while parsing D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java
```

>> the processing errors as part of the generated pmd.xml file

I see no such file anywhere even though I did not trust the default setting for `renderProcessingErrors`

```xml
            <plugin> 
                   <groupId>org.apache.maven.plugins</groupId> 
                   <artifactId>maven-pmd-plugin</artifactId> 
                   <version>3.9.0</version> 
                   <configuration> 
                       <targetJdk>${javac.target}</targetJdk> 
                       <printFailingErrors>true</printFailingErrors> 
                       <skipPmdError>false</skipPmdError> 
                       <renderProcessingErrors>true</renderProcessingErrors>
                       <excludeRoots> 
                           <excludeRoot>target/generated-sources/java</excludeRoot> 
                       </excludeRoots> 
                   </configuration> 
                   <dependencies> 
                       <dependency> 
                          <groupId>net.sourceforge.pmd</groupId> 
                          <artifactId>pmd-core</artifactId> 
                          <version>${pmd.version}</version> 
                       </dependency> 
                       <dependency> 
                          <groupId>net.sourceforge.pmd</groupId> 
                          <artifactId>pmd-java</artifactId> 
                          <version>${pmd.version}</version> 
                       </dependency> 
                   </dependencies> 
               </plugin>
```

After running with _-X_ option enabled found the following:
```
[ERROR] D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java: Error while processing D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java
net.sourceforge.pmd.PMDException: Error while processing D:\Projects\apache\mina-sshd\sshd-core\src\main\java\org\apache\sshd\common\file\util\BasePath.java
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:101)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:50)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:75)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:23)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinitionSimple.toString(JavaTypeDefinitionSimple.java:249)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinitionUpper.toString(JavaTypeDefinitionUpper.java:116)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodType.toString(MethodType.java:78)
	at java.lang.String.valueOf(String.java:2994)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.selectMethodsThirdPhase(MethodTypeResolution.java:358)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.getBestMethodReturnType(MethodTypeResolution.java:384)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:434)
	at net.sourceforge.pmd.lang.java.ast.ASTName.jjtAccept(ASTName.java:35)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:314)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:1019)
	at net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix.jjtAccept(ASTPrimaryPrefix.java:42)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:855)
	at net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression.jjtAccept(ASTPrimaryExpression.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:230)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:715)
	at net.sourceforge.pmd.lang.java.ast.ASTExpression.jjtAccept(ASTExpression.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:422)
	at net.sourceforge.pmd.lang.java.ast.ASTReturnStatement.jjtAccept(ASTReturnStatement.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:350)
	at net.sourceforge.pmd.lang.java.ast.ASTStatement.jjtAccept(ASTStatement.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:362)
	at net.sourceforge.pmd.lang.java.ast.ASTBlockStatement.jjtAccept(ASTBlockStatement.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:358)
	at net.sourceforge.pmd.lang.java.ast.ASTBlock.jjtAccept(ASTBlock.java:25)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:182)
	at net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration.jjtAccept(ASTMethodDeclaration.java:31)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:46)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBodyDeclaration.jjtAccept(ASTClassOrInterfaceBodyDeclaration.java:41)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:42)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBody.jjtAccept(ASTClassOrInterfaceBody.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:18)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:339)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration.jjtAccept(ASTClassOrInterfaceDeclaration.java:39)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:158)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:251)
	at net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration.jjtAccept(ASTTypeDeclaration.java:35)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:138)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:222)
	at net.sourceforge.pmd.lang.java.ast.ASTCompilationUnit.jjtAccept(ASTCompilationUnit.java:41)
	at net.sourceforge.pmd.lang.java.typeresolution.TypeResolutionFacade.initializeWith(TypeResolutionFacade.java:18)
	at net.sourceforge.pmd.lang.java.AbstractJavaHandler$5.start(AbstractJavaHandler.java:97)
	at net.sourceforge.pmd.SourceCodeProcessor.usesTypeResolution(SourceCodeProcessor.java:149)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:177)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
	... 9 more

D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java: Error while parsing D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java
net.sourceforge.pmd.PMDException: Error while parsing D:\Projects\apache\mina-sshd\sshd-core\src\test\java\org\apache\sshd\server\jaas\JaasPasswordAuthenticatorTest.java
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:98)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:50)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:75)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:23)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Line 70, Column 9: Cannot use '<>' with anonymous inner classes when running in JDK inferior to 9 mode!
	at net.sourceforge.pmd.lang.java.ast.JavaParser.throwParseException(JavaParser.java:23)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.checkForBadAnonymousDiamondUsage(JavaParser.java:139)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.AllocationExpression(JavaParser.java:4519)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.PrimaryPrefix(JavaParser.java:3922)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.PrimaryExpression(JavaParser.java:3735)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.PostfixExpression(JavaParser.java:3595)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.UnaryExpressionNotPlusMinus(JavaParser.java:3557)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.UnaryExpression(JavaParser.java:3410)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.MultiplicativeExpression(JavaParser.java:3304)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.AdditiveExpression(JavaParser.java:3251)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ShiftExpression(JavaParser.java:3196)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.RelationalExpression(JavaParser.java:3135)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.InstanceOfExpression(JavaParser.java:3099)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.EqualityExpression(JavaParser.java:3046)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.AndExpression(JavaParser.java:3006)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ExclusiveOrExpression(JavaParser.java:2966)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.InclusiveOrExpression(JavaParser.java:2926)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ConditionalAndExpression(JavaParser.java:2886)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ConditionalOrExpression(JavaParser.java:2846)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ConditionalExpression(JavaParser.java:2807)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Expression(JavaParser.java:2666)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.VariableInitializer(JavaParser.java:1462)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.VariableDeclarator(JavaParser.java:1337)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.LocalVariableDeclaration(JavaParser.java:4911)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:4834)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:4795)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.MethodDeclaration(JavaParser.java:1618)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBodyDeclaration(JavaParser.java:1241)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBody(JavaParser.java:1196)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceDeclaration(JavaParser.java:645)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.TypeDeclaration(JavaParser.java:534)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.CompilationUnit(JavaParser.java:271)
	at net.sourceforge.pmd.lang.java.AbstractJavaParser.parse(AbstractJavaParser.java:57)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:110)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:173)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
	... 9 more
```

Seems that:

* _BasePath_ - some kind of NPE
```
Caused by: java.lang.NullPointerException
	at net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinitionSimple.toString(JavaTypeDefinitionSimple.java:249)
```
* _JaasPasswordAuthenticatorTest_ - wrongly detects "anonymous" diamond operator at line 70 - I can assure you the code is 100% Java-8 compliant + I am not using Java 9 anywhere - not even to compile it

Here is the "offending" snippet (you can see the full code in the GIT repo):
```java
    @Before
    public void setUp() {
        Configuration config = new Configuration() {
            @Override
            public AppConfigurationEntry[] getAppConfigurationEntry(String name) {
                return new AppConfigurationEntry[]{
                    new AppConfigurationEntry(DummyLoginModule.class.getName(),
                            AppConfigurationEntry.LoginModuleControlFlag.REQUIRED,
                            new HashMap<>())
                };
            }

            @Override
            public void refresh() {
                // ignored
            }
        };
        Configuration.setConfiguration(config);
    }
```

Which is reported as:

```
Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Line 70, Column 9: Cannot use '<>' with anonymous inner classes when running in JDK inferior to 9 mode!
	at net.sourceforge.pmd.lang.java.ast.JavaParser.throwParseException(JavaParser.java:23)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.checkForBadAnonymousDiamondUsage(JavaParser.java:139)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.AllocationExpression(JavaParser.java:4519)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.PrimaryPrefix(JavaParser.java:3922)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.PrimaryExpression(JavaParser.java:3735)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.PostfixExpression(JavaParser.java:3595)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.UnaryExpressionNotPlusMinus(JavaParser.java:3557)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.UnaryExpression(JavaParser.java:3410)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.MultiplicativeExpression(JavaParser.java:3304)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.AdditiveExpression(JavaParser.java:3251)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ShiftExpression(JavaParser.java:3196)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.RelationalExpression(JavaParser.java:3135)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.InstanceOfExpression(JavaParser.java:3099)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.EqualityExpression(JavaParser.java:3046)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.AndExpression(JavaParser.java:3006)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ExclusiveOrExpression(JavaParser.java:2966)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.InclusiveOrExpression(JavaParser.java:2926)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ConditionalAndExpression(JavaParser.java:2886)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ConditionalOrExpression(JavaParser.java:2846)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ConditionalExpression(JavaParser.java:2807)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Expression(JavaParser.java:2666)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.VariableInitializer(JavaParser.java:1462)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.VariableDeclarator(JavaParser.java:1337)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.LocalVariableDeclaration(JavaParser.java:4911)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:4834)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:4795)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.MethodDeclaration(JavaParser.java:1618)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBodyDeclaration(JavaParser.java:1241)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBody(JavaParser.java:1196)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceDeclaration(JavaParser.java:645)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.TypeDeclaration(JavaParser.java:534)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.CompilationUnit(JavaParser.java:271)
	at net.sourceforge.pmd.lang.java.AbstractJavaParser.parse(AbstractJavaParser.java:57)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:110)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:173)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
```

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -58,6 +58,9 @@ we have measured up to 10% improvements during Type Resolution, Symbol Table ana
     *   [#988](https://github.com/pmd/pmd/issues/988): \[core] FileNotFoundException for missing classes directory with analysis cache enabled
 *   documentation
     *   [#994](https://github.com/pmd/pmd/issues/994): \[doc] Delete duplicate page contributing.md on the website
+*   java
+    *   [#894](https://github.com/pmd/pmd/issues/894): \[java] Maven PMD plugin fails to process some files without any explanation
+    *   [#899](https://github.com/pmd/pmd/issues/899): \[java] JavaTypeDefinitionSimple.toString can cause NPEs
 *   java-bestpractices
     *   [#370](https://github.com/pmd/pmd/issues/370): \[java] GuardLogStatementJavaUtil not considering lambdas
     *   [#558](https://github.com/pmd/pmd/issues/558): \[java] ProperLogger Warnings for enums

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java
@@ -83,6 +83,7 @@ public class ClassNamingConventionsRule extends AbstractJavaRule {
                 if (!((ASTInitializer) decl.getDeclarationNode()).isStatic()) {
                     return false;
                 }
+                break;
 
             default:
                 break;

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinitionSimple.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinitionSimple.java
@@ -246,6 +246,11 @@ import java.util.logging.Logger;
                 .append(", definitionType=").append(getDefinitionType())
                 .append(", genericArgs=[");
         
+        // Forcefully resolve all generic types
+        for (int i = 0; i < genericArgs.size(); i++) {
+            getGenericType(i);
+        }
+        
         for (final JavaTypeDefinition jtd : genericArgs) {
             sb.append(jtd.shallowString()).append(", ");
         }

--- a/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/ClassTypeResolverTest.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/ClassTypeResolverTest.java
@@ -92,6 +92,7 @@ import net.sourceforge.pmd.typeresolution.testdata.FieldAccessSuper;
 import net.sourceforge.pmd.typeresolution.testdata.GenericMethodsImplicit;
 import net.sourceforge.pmd.typeresolution.testdata.GenericsArrays;
 import net.sourceforge.pmd.typeresolution.testdata.InnerClass;
+import net.sourceforge.pmd.typeresolution.testdata.JavaTypeDefinitionToStringNPE;
 import net.sourceforge.pmd.typeresolution.testdata.Literals;
 import net.sourceforge.pmd.typeresolution.testdata.MethodAccessibility;
 import net.sourceforge.pmd.typeresolution.testdata.MethodFirstPhase;
@@ -177,6 +178,15 @@ public class ClassTypeResolverTest {
         Class<?> inner = acu.getFirstDescendantOfType(ASTAllocationExpression.class)
                 .getFirstDescendantOfType(ASTClassOrInterfaceType.class).getType();
         assertEquals("net.sourceforge.pmd.typeresolution.testdata.EnumWithAnonymousInnerClass$2", inner.getName());
+    }
+    
+    /**
+     * See bug #899 toString causes NPE
+     */
+    @Test
+    public void testNPEInJavaTypeDefinitionToString() {
+        // Just parsing this file throws a NPE
+        parseAndTypeResolveForClass(JavaTypeDefinitionToStringNPE.class, "1.8");
     }
 
     @Test

--- a/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/testdata/JavaTypeDefinitionToStringNPE.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/testdata/JavaTypeDefinitionToStringNPE.java
@@ -1,0 +1,23 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.typeresolution.testdata;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class JavaTypeDefinitionToStringNPE<A extends JavaTypeDefinitionToStringNPE.SelfReferringType<A>> {
+
+    Map<TypeLink<?, ? super A>, Collection<?>> contents = new HashMap<>();
+
+    public static class TypeLink<U, T> {}
+
+    public static class SelfReferringType<T extends SelfReferringType<T>> {}
+
+    public final void putNull(TypeLink<?, ? super A> field) {
+        contents.put(field, Collections.singleton(null));
+    }
+}

--- a/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/testdata/JavaTypeDefinitionToStringNPE.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/testdata/JavaTypeDefinitionToStringNPE.java
@@ -13,9 +13,9 @@ public class JavaTypeDefinitionToStringNPE<A extends JavaTypeDefinitionToStringN
 
     Map<TypeLink<?, ? super A>, Collection<?>> contents = new HashMap<>();
 
-    public static class TypeLink<U, T> {}
+    public static class TypeLink<U, T> { }
 
-    public static class SelfReferringType<T extends SelfReferringType<T>> {}
+    public static class SelfReferringType<T extends SelfReferringType<T>> { }
 
     public final void putNull(TypeLink<?, ? super A> field) {
         contents.put(field, Collections.singleton(null));

